src/server/sql/reading/calculate_compare.sql

```diff
@@ -0,0 +1,106 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+	The compare chart needs three pieces of raw information for each meter:
+
+	1. How much was used during this time period?
+	2. How much was used during the previous time period?
+	3. How much was used up to this point in the previous time period
+ */
+
+/*
+This shouldn't ever be looking at more than a few weeks of data, so we don't need to deal with compression.
+ */
+CREATE FUNCTION compare_readings(
+	meter_ids INTEGER[],
+	current_period_start TIMESTAMP,
+	current_period_end TIMESTAMP,
+	period_duration INTERVAL
+)
+	RETURNS TABLE(meter_id INTEGER, current_use REAL, prev_use_total REAL, prev_use_for_current REAL)
+AS $$
+DECLARE
+	prev_period_start TIMESTAMP;
+	curr_period_duration INTERVAL; -- How far into the current period we are
+	prev_period_partial_end TIMESTAMP;
+BEGIN
+	curr_period_duration := current_period_end - current_period_start;
+	ASSERT curr_period_duration <= period_duration, 'Current period is too long';
+	prev_period_start := current_period_start - period_duration;
+	prev_period_partial_end := prev_period_start + curr_period_duration;
+
+	RETURN QUERY
+	WITH
+	current_period AS (
+		SELECT
+			m.id AS meter_id,
+			SUM(r.reading) AS reading
+		FROM readings r
+		INNER JOIN unnest(meter_ids) m(id) ON r.meter_id = m.id
+		WHERE r.start_timestamp >= current_period_start AND r.end_timestamp <= current_period_end
+		GROUP BY m.id
+	),
+	prev_period AS (
+		SELECT
+			m.id AS meter_id,
+			SUM(r.reading) AS reading
+		FROM readings r
+			INNER JOIN unnest(meter_ids) m(id) ON r.meter_id = m.id
+		WHERE r.start_timestamp >= prev_period_start AND r.end_timestamp <= current_period_start -- prev_end == curr_start
```

Comment:
prev_end = curr_start only when curr_period_duration = period_duration. I think that both the comment and logic should be replaced but it may occur so rarely that they are not equal that it won't make sense to implement. It is just one more addition though so...

Reply:
That's not actually true. The start of the current period will always be the end of the one before it, but the end of the current one isn't always (and is usually not) `prev_end + period_duration` since the current period isn't usually over (instead it's `prev_end + curr_duration`).

But we can assume that the current one starts where the previous one ends.

I might have some variable names slightly wrong; I haven't looked at this in months.

Reply:
I see it now. Thank you for the insight!